File: backend/src/core/forms/application/calculated-field.engine.ts
import { Injectable } from '@nestjs/common';
import { FormFieldDefinition } from '@agroerp/shared';

const SAFE_EXPRESSION = /^[\d\s+\-*/().{}a-zA-Z_]+$/;

@Injectable()
export class CalculatedFieldEngine {
  resolve(
    fields: FormFieldDefinition[],
    data: Record<string, unknown>,
  ): Record<string, unknown> {
    const resolved = { ...data };

    for (const field of fields) {
      if (field.type !== 'calculated' || !field.calculate) continue;

      const missing = field.calculate.dependsOn.some(
        (key) => resolved[key] === undefined || resolved[key] === null,
      );
      if (missing) continue;

      try {
        resolved[field.key] = this.evaluateExpression(
          field.calculate.expression,
          resolved,
        );
      } catch {
        // Skip invalid calculated fields; validation will catch if required
      }
    }

    return resolved;
  }

  private evaluateExpression(
    expression: string,
    data: Record<string, unknown>,
  ): number {
    let expr = expression;
    for (const [key, value] of Object.entries(data)) {
      const token = `{${key}}`;
      if (expr.includes(token)) {
        const num = Number(value);
        if (Number.isNaN(num)) {
          throw new Error(`Calculated field dependency "${key}" is not numeric`);
        }
        expr = expr.split(token).join(String(num));
      }
    }

    if (!SAFE_EXPRESSION.test(expr)) {
      throw new Error('Unsafe calculated expression');
    }

    // eslint-disable-next-line no-new-func
    const result = Function(`"use strict"; return (${expr});`)();
    if (typeof result !== 'number' || Number.isNaN(result)) {
      throw new Error('Calculated expression did not return a number');
    }
    return result;
  }
}

File: mobile-android/app/src/main/java/com/agroerp/domain/engine/CalculatedFieldEngine.kt
package com.agroerp.domain.engine

import com.agroerp.domain.model.FormField
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CalculatedFieldEngine @Inject constructor() {

    private val safeExpression = Regex("""^[\d\s+\-*/().{}a-zA-Z_]+$""")

    fun resolve(
        fields: List<FormField>,
        data: Map<String, Any?>,
    ): Map<String, Any?> {
        val resolved = data.toMutableMap()
        for (field in fields) {
            if (field.type != "calculated" || field.calculate == null) continue
            val config = field.calculate
            if (config.dependsOn.any { resolved[it] == null }) continue
            try {
                resolved[field.key] = evaluateExpression(config.expression, resolved)
            } catch (_: Exception) {
                // Skip invalid calculated values
            }
        }
        return resolved
    }

    private fun evaluateExpression(
        expression: String,
        data: Map<String, Any?>,
    ): Double {
        var expr = expression
        for ((key, value) in data) {
            val token = "{$key}"
            if (expr.contains(token)) {
                val num = (value as? Number)?.toDouble()
                    ?: value.toString().toDoubleOrNull()
                    ?: throw IllegalArgumentException("Non-numeric dependency: $key")
                expr = expr.replace(token, num.toString())
            }
        }
        if (!safeExpression.matches(expr)) {
            throw IllegalArgumentException("Unsafe expression")
        }
        return ExpressionEvaluator.eval(expr)
    }
}

/** Minimal safe arithmetic evaluator (no Function/JS). */
private object ExpressionEvaluator {
    fun eval(expr: String): Double {
        return Parser(expr.replace(" ", "")).parse()
    }

    private class Parser(private val input: String) {
        private var pos = 0

        fun parse(): Double = parseExpression()

        private fun parseExpression(): Double {
            var result = parseTerm()
            while (pos < input.length) {
                when (input[pos]) {
                    '+' -> { pos++; result += parseTerm() }
                    '-' -> { pos++; result -= parseTerm() }
                    else -> break
                }
            }
            return result
        }

        private fun parseTerm(): Double {
            var result = parseFactor()
            while (pos < input.length) {
                when (input[pos]) {
                    '*' -> { pos++; result *= parseFactor() }
                    '/' -> { pos++; result /= parseFactor() }
                    else -> break
                }
            }
            return result
        }

        private fun parseFactor(): Double {
            if (input[pos] == '(') {
                pos++
                val result = parseExpression()
                if (pos < input.length && input[pos] == ')') pos++
                return result
            }
            val start = pos
            if (input[pos] == '-') pos++
            while (pos < input.length && (input[pos].isDigit() || input[pos] == '.')) pos++
            return input.substring(start, pos).toDouble()
        }
    }
}
